### DISCUSSION TITLE
Renovate with GitLab Project Access Tokens?

### DISCUSSION THREAD

**Discussion by chris-dura** [renovatebot/renovate#9167] [Request Help] [answered]
Our IT department is bulking at creating a renovatebot GitLab User, since that GitLab user isn't technically "owned" by anyone, and would not be tied to a particular individual.

As a temporary workaround, I've just be letting the bot use my Personal Access Token; however, as a matter of practice, that doesn't make sense, and isn't very secure.

GitLab does provide "Project Access Tokens" that enable access scoped to one particular project/repo. (The GitLab devs are also working on a "Group Access Token" feature, which, I would imagine scopes the token to the Group-level.)

In a typical CI/CD setup, I'm providing the bot a single **Personal** Access Token, and then giving the bot a list of projects to go renovate. However, assuming **Project** Access Tokens would work the same way (just on single projects) --

Does anyone have any brilliant idea on how I could approach giving the bot a _list_ of access tokens to use, instead of just the one, and which project to use them on?

**Comment by rarkins** [comment, marked as answer]
I don't think there's any way to get this to work today as-is, without writing your own custom wrapper script and running Renovate 1 repo at a time.

- Would project access tokens be long-lived (generated once, used many times) or generated every single time Renovate needs to run on a repo?
- If they're long-lived, who would generate them and how would they be stored and accessed? e.g. generated manually by yourself and added to Renovate's config?

**Reply by chris-dura** [reply]
@rarkins --

1. Yes, the Project Tokens would be long-lived, the GitLab docs indicate that creating a Project Access Token automatically creates a bot user and gives them Maintainer-level permissions to the repo. I _just_ discovered Project Access Tokens, but the way the docs read, I _think_ the idea is if you have an "app" (in this case, renovate) you can give it a Project Token, and the "app" can use that to authenticate to the API. I can't tell if it works _exactly_ like Personal Tokens, but in the API docs, it implies it does, at least as far as API access for that one particular project 🤷 
https://docs.gitlab.com/ee/user/project/settings/project_access_tokens.html#project-bot-users
https://docs.gitlab.com/ee/api/README.html#personalproject-access-tokens

2. Myself (or anyone with a project they wanted to renovate) would generate the Tokens. I also think they'd be stored as masked/protected CI/CD variables (which is how I would store the one Personal Access Token in a typical setup, too, I think)

If Project Tokens _are_ functionally used the same as Personal Tokens, the best klugy workaround I can come up with off the top of my head is to maybe run renovatebot MANY times with MANY configs (one config per project). And, I could probably do this with either MANY pipeline jobs... or even separate branches... not sure which would be "better" 🤷

**Reply by rarkins** [reply]
If you ran a pipeline per-repo (i.e. each repo runs Renovate on itself) then that might be one way to do it.

**Reply by chris-dura** [reply]
Yeah, that would probably be a more scalable solution.

**Comment by viceice** [comment]
Can you try this:

**config.js**:
```js
modules.exports = {
  token: '<PERSONAL_TOKEN>',
  repositories: [
    {
      repository: 'group/repo',
      token: '<PROJECT_TOKEN>',
    }
  ],
};
```

So mayb this helps renovate to use the personal token for discovery and using project token for repo processing.

**Reply by rarkins** [reply]
JSON needs fixing. Never tried that, I think we pass the token to Gitlab during initPlatform

**Reply by viceice** [reply]
right, doesn't work.

**Comment by rikard-helgegren** [comment]
I used a Group Access Token instead of a Personal Access Token, which worked great
